master/internal/stream/experiments.go

```diff
@@ -0,0 +1,277 @@
+	ws := stream.WhereSince{Since: 0}
+	if len(spec.ExperimentIds) > 0 {
+		ws.Include("e.id in (?)", bun.In(spec.ExperimentIds))
+	}
+	q = ws.Apply(q)
+
+	var exist []int64
+	err = q.Scan(ctx, &exist)
+	if err != nil && errors.Cause(err) != sql.ErrNoRows {
+		log.Errorf("error: %v\n", err)
+		return nil, err
+	}
+
+	// step 2: figure out what was missing and what has appeared
+	missing, appeared, err := stream.ProcessKnown(known, exist)
+	if err != nil {
+		return nil, err
+	}
+
+	// step 3: hydrate appeared IDs into full TrialMsgs
+	var expMsgs []*ExperimentMsg
+	if len(appeared) > 0 {
+		query := getExperimentMsgsWithWorkspaceID(expMsgs).
+			Where("experiment_msg.id in (?)", bun.In(appeared))
+		query = permFilter(query, accessMap, accessScopes)
+		err := query.Scan(ctx, &expMsgs)
+		if err != nil && errors.Cause(err) != sql.ErrNoRows {
+			log.Errorf("error: %v\n", err)
+			return nil, err
+		}
+	}
+
+	// step 4: emit deletions and updates to the client
+	out = append(out, stream.DeleteMsg{
+		Key:     ExperimentsDeleteKey,
+		Deleted: missing,
+	})
+	for _, msg := range expMsgs {
+		out = append(out, msg.UpsertMsg())
+	}
+	return out, nil
+}
+
+// ExperimentCollectSubscriptionModMsgs scrapes the database when a
+// user submits a new TrialSubscriptionSpec for initial matches.
+func ExperimentCollectSubscriptionModMsgs(ctx context.Context, addSpec ExperimentSubscriptionSpec) (
+	[]interface{}, error,
+) {
+	if len(addSpec.ExperimentIds) == 0 {
+		return nil, nil
+	}
+	var expMsgs []*ExperimentMsg
+	q := getExperimentMsgsWithWorkspaceID(expMsgs)
+
+	// Use WhereSince to build a complex WHERE clause.
+	ws := stream.WhereSince{Since: addSpec.Since}
+	if len(addSpec.ExperimentIds) > 0 {
+		ws.Include("experiment_id in (?)", bun.In(addSpec.ExperimentIds))
+	}
+	q = ws.Apply(q)
+
+	err := q.Scan(ctx)
+	if err != nil && errors.Cause(err) != sql.ErrNoRows {
+		log.Errorf("error: %v\n", err)
+		return nil, err
+	}
+
+	var out []interface{}
+	for _, msg := range expMsgs {
+		out = append(out, msg.UpsertMsg())
+	}
+	return out, nil
+}
+
+// ExperimentFilterMaker tracks the trial and experiment id's that are to be filtered for.
+type ExperimentFilterMaker struct {
+	ExperimentIds map[int]bool
+}
+
+// NewExperimentFilterMaker creates a new FilterMaker.
+func NewExperimentFilterMaker() FilterMaker[*ExperimentMsg, ExperimentSubscriptionSpec] {
+	return &ExperimentFilterMaker{make(map[int]bool)}
+}
+
+// AddSpec adds ExperimentIds specified in ExpSubscriptionSpec.
+func (e *ExperimentFilterMaker) AddSpec(spec ExperimentSubscriptionSpec) {
+	for _, id := range spec.ExperimentIds {
+		e.ExperimentIds[id] = true
+	}
+}
+
+// DropSpec removes TrialIds and ExperimentIds specified in ExpSubscriptionSpec.
+func (e *ExperimentFilterMaker) DropSpec(spec ExperimentSubscriptionSpec) {
+	for _, id := range spec.ExperimentIds {
+		delete(e.ExperimentIds, id)
+	}
+}
+
+// MakeFilter returns a function that determines if a TrialMsg based on
```

Comment:
Should be `ExperimentMsg`, running a find and replace on "Trial" w/ "Experiment" should resolve these (just be sure to deserve the original case)